source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1018,6 +1019,172 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
+      if (callback != nullptr) {
+        (*callback)();
+      }
+    }
+    pending_clusters_.erase(name);
+    if (!callbacks_handle_cleanup_posted_ && pending_clusters_.empty()) {
+      parent_.thread_local_dispatcher_.post([this] {
+        // Something might got added in the meantime, so check the
+        // map again.
+        if (pending_clusters_.empty()) {
+          resetCallbacks();
+        }
+        callbacks_handle_cleanup_posted_ = false;
+      });
+      callbacks_handle_cleanup_posted_ = true;
+    }
+  });
+  callbacks_ = std::make_unique<clusterCallbacks>(cb);
+  callbacks_handle_ = parent_.parent_.addThreadLocalClusterUpdateCallbacks(*callbacks_);
+}
+
+ClusterManagerImpl::ClusterDiscoveryManager::Pair
+ClusterManagerImpl::ClusterDiscoveryManager::addCallback(
+    const std::string& name, ClusterDiscoveryCallbackWeakPtr weak_callback) {
+  auto& callbacks_deque = pending_clusters_[name];
+  callbacks_deque.emplace_back(weak_callback);
+  auto handle =
+      std::make_unique<ClusterDiscoveryCallbackHandleImpl>(*this, std::move(weak_callback), name);
+  auto discovery_in_progress = (callbacks_deque.size() > 1);
+  return {std::move(handle), discovery_in_progress};
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::erase(const std::string& name,
+                                                        ClusterDiscoveryCallbackWeakPtr cb) {
+  const bool drop_deque = eraseFromDeque(name, cb);
+  if (drop_deque) {
+    pending_clusters_.erase(name);
+  }
+  if (pending_clusters_.empty()) {
+    resetCallbacks();
+  }
+}
+
+bool ClusterManagerImpl::ClusterDiscoveryManager::eraseFromDeque(
+    const std::string& name, ClusterDiscoveryCallbackWeakPtr cb) {
+  auto it = pending_clusters_.find(name);
+  if (it == pending_clusters_.end()) {
+    return false;
+  }
+  auto& deque = it->second;
+  // Could use std::erase_if, but it's only in C++20.
+  auto it2 = std::remove_if(deque.begin(), deque.end(),
+                            [cb](ClusterDiscoveryCallbackWeakPtr const& weak_ptr) {
+                              if (cb.owner_before(weak_ptr))
+                                return false;
+                              if (weak_ptr.owner_before(cb))
+                                return false;
+                              return true;
+                            });
+  deque.erase(it2, deque.end());
+  return deque.empty();
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::resetCallbacks() {
+  callbacks_handle_.reset();
+  callbacks_.reset();
+}
+
+ClusterDiscoveryCallbackHandlePtr
+ClusterManagerImpl::requestOnDemandClusterDiscovery(CdsApiWeakPtr weak_cds, const std::string& name,
+                                                    ClusterDiscoveryCallbackWeakPtr weak_callback) {
+  ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
+
+  cluster_manager.cdm_.ensureCallbacksAreInstalled();
+
+  auto [handle, discovery_in_progress] = cluster_manager.cdm_.addCallback(name, weak_callback);
```

Comment:
This looks like it will miss discovery requests for `name` initiated in other threads, is this intentional? Could you add a comment then?

Reply:
Will add more comments, sure.

Note that `addCallback` will add a callback anyway. But if the discovery for `name` was already initiated in this thread, then we don't need to do anything else, just add the callback and wait for it to be fired. If the discovery for `name` was initiated in some other worker thread, then we have no way of knowing that in this worker thread, so we try to initiate the discovery anyway. Whether the discovery will actually be initiated is something that main thread will figure out (see the code where we check `pending_cluster_creations_` in main thread).